libp2p/protocols/kademlia/kademlia.nim

```diff
@@ -68,6 +76,61 @@ proc waitRepliesOrTimeouts(
 
   return (receivedReplies, failedPeers)
 
+proc dispatchPutVal(
+    kad: KadDHT, peer: PeerId, entry: ValidatedEntry
+): Future[void] {.
+    async: (raises: [CancelledError, DialFailedError, ValueError, LPStreamError])
+.} =
+  let conn = await kad.switch.dial(peer, KadCodec)
+
+  defer:
+    await conn.close()
+
+  let msg = Message(
+    msgType: MessageType.putValue,
+    record: some(Record(key: some(entry.key.data), value: some(entry.val.data))),
+  )
+
+  await conn.writeLp(msg.encode().buffer)
+
+  let reply = Message.decode(await conn.readLp(MaxMsgSize)).tryGet()
+
+  if reply.msgType != MessageType.putValue:
+    raise newException(ValueError, "unexpected message type in reply: " & $reply)
+
+proc putVal*(
+    kad: KadDHT, key: keys.Key, val: EntryVal, replic: int
+): Future[Result[void, string]] {.async: (raises: [CancelledError]), gcsafe.} =
+  let cand = EntryCandidate(key: EntryKey(data: key.getBytes()), val: val)
+  if not kad.entryValidator.validate(cand):
+    return err("invalid key/val pair")
+  let ts = TimeStamp(ts: $times.now().utc)
+  try:
+    let others: seq[RecordVal] =
```

Comment:
This still logs an error in tests: `ERR 2025-08-05 11:12:03.195+02:00 could not send find_node to peer           topics="kad-dht" tid=214810 peerId=12D*W9z11w err="failed new dial: internalConnect can't dial self!"`